datastore/src/main/java/io/spine/server/storage/datastore/query/DsLookupByQueries.java

```diff
@@ -0,0 +1,205 @@
+    private final Optional<Filter> ancestorFilter;
+
+    /**
+     * A converter from {@link Entity} to {@code <R>} instances.
+     *
+     * <p>Initialized upon fetching the entities from the Datastore.
+     */
+    private @MonotonicNonNull ToRecords<R> transformer = null;
+
+    /**
+     * Creates a new lookup for the passed {@code RecordQuery}.
+     *
+     * @param datastore
+     *         Datastore connector
+     * @param query
+     *         a query to create this lookup for
+     * @param columnAdapter
+     *         an adapter of {@code RecordQuery} parameter values to Datastore-native types
+     * @param spec
+     *         Entity specification of the queried records
+     */
+    DsLookupByQueries(DatastoreMedium datastore,
+                      RecordQuery<I, R> query,
+                      FilterAdapter columnAdapter,
+                      DsEntitySpec<I, R> spec) {
+        super(query, columnAdapter, spec);
+        this.datastore = datastore;
+        this.ancestorFilter = ancestorFilter(query, datastore);
+    }
+
+    private Optional<Filter> ancestorFilter(RecordQuery<I, R> query, DatastoreMedium datastore) {
+        Optional<Filter> result = spec().layout()
+                                        .ancestorFilter(query, datastore);
+        return result;
+    }
+
+    @Override
+    IntermediateResult fetchFromDatastore() {
+        ImmutableList<Entity> rawEntities = findByPredicates(query());
+        return new IntermediateResult(rawEntities);
+    }
+
+    @Override
+    Iterable<R> toRecords(IntermediateResult result) {
+        checkNotNull(transformer,
+                     "In-memory transformer `Datastore Entity`->`Stored record` isn't set.");
+        Iterable<R> records = transformer.apply(result);
+        return records;
+    }
+
+    private ImmutableList<Entity> findByPredicates(RecordQuery<?, R> query) {
+        ImmutableList<Entity> results;
+        List<StructuredQuery<Entity>> queries = split(query);
+        if (queries.size() == 1) {
+            results = runSingleQuery(queries.get(0));
+            transformer = new ConvertAsIs<>(recordType(), mask());
+        } else {
+            results = readAndJoin(queries);
+            transformer = new SortAndLimit<>(recordType(), mask(), sorting(), limit());
+        }
+
+        return results;
+    }
+
+    private List<StructuredQuery<Entity>> split(RecordQuery<?, R> query) {
+        QueryPredicate<R> rootPredicate = query.subject()
+                                               .predicate();
+        Kind kind = spec().kind();
+        if (rootPredicate.isEmpty()) {
+            StructuredQuery<Entity> result = new QueryWithFilter(query, kind).withNoFilter();
+            return ImmutableList.of(result);
+        }
+
+        List<StructuredQuery<Entity>> queries = toDatastoreFilters(rootPredicate)
+                .stream()
+                .map(new QueryWithFilter(query, kind))
+                .collect(toList());
```

Comment:
Is there a reason for returning `List` rather than `ImmutableList` in this particular case?